### PR TITLE
Option rework preparations

### PR DESCRIPTION
**Pull request by BlackYps**
The only functional change here is that it is now possible to specify a map style and the symmetry settings at the same time.

The bulk of the changes is a refactoring of how the options are nested and the classes are named.

### CODE REVIEW

File: generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java [hunk 1]
```diff
@@ -5,14 +5,12 @@
...
-    private VisibilityOptions visibilityOptions;
-    @CommandLine.ArgGroup(order = 1, heading = "Options to manually tune map generation%n", exclusive = false)
-    private ParameterOptions parameterOptions;
+public class StyleOptions {
+    @CommandLine.ArgGroup(order = 2, heading = "Options to create a custom map style%n", exclusive = false)
+    private CustomStyleOptions customStyleOptions;
     private MapStyle mapStyle;
 
-    @CommandLine.Option(names = "--style", order = 1, description = "Style for the generated map. Values: ${COMPLETION-CANDIDATES}")
+    @CommandLine.Option(names = "--style", order = 6, description = "Style for the generated map. Values: ${COMPLETION-CANDIDATES}")
```
Comment (Sheikah45) [inline generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java]: style should still have order 1 so it is at the top

File: generator/src/main/java/com/faforever/neroxis/generator/cli/BasicOptions.java [hunk 2]
```diff
@@ -5,20 +5,22 @@
...
+    @CommandLine.Option(names = "--spawn-count", order = 5, defaultValue = "6", description = "Spawn count for the generated map", showDefaultValue = CommandLine.Help.Visibility.ALWAYS)
     private Integer spawnCount;
-    @CommandLine.Option(names = "--num-teams", order = 4, defaultValue = "2", description = "Number of teams for the generated map (0 is no teams asymmetric)", showDefaultValue = CommandLine.Help.Visibility.ALWAYS)
+    @CommandLine.Option(names = "--num-teams", order = 6, defaultValue = "2", description = "Number of teams for the generated map (0 is no teams asymmetric)", showDefaultValue = CommandLine.Help.Visibility.ALWAYS)
     private Integer numTeams;
     private Integer mapSize;
 
-    @CommandLine.Option(names = "--map-size", order = 2, defaultValue = "512", description = "Generated map size, can be specified in oGrids (e.g 512) or km (e.g 10km)", showDefaultValue = CommandLine.Help.Visibility.ALWAYS)
+    @CommandLine.Option(names = "--map-size", order = 4, defaultValue = "512", description = "Generated map size, can be specified in oGrids (e.g 512) or km (e.g 10km)", showDefaultValue = CommandLine.Help.Visibility.ALWAYS)
```
Comment (Sheikah45) [inline generator/src/main/java/com/faforever/neroxis/generator/cli/BasicOptions.java:23]: why did the order values change? Order is relative to the argGroup

**Review by Sheikah45 — CHANGES_REQUESTED** [review]

File: generator/src/main/java/com/faforever/neroxis/generator/cli/BasicOptions.java (hunk 2 above)
Comment (BlackYps) [reply generator/src/main/java/com/faforever/neroxis/generator/cli/BasicOptions.java:23]: To be honest I didn't understand how the order of groups and options relates. I got the weird behavior that the heading of a group was at the very bottom while its contents were higher up. I didn't even think that was possible. So I fiddled with the values until at least the group headings were not detached from their content

File: generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java (hunk 1 above)
Comment (BlackYps) [reply generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java]: I set it to 1, but it is still not at the top

Comment (Sheikah45) [reply generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java]: is it at least at the top of the group?

Comment (BlackYps) [reply generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java]: This is the current output
```
Usage: generate [-hV] [--debug] [--visualize] [--map-name=<mapName>]
                [--num-to-generate=<numToGenerate>] [--out-path=<outputPath>]
                [--preview-path=<previewFolder>] [[--seed=<seed>]
                [--spawn-count=<spawnCount>] [--num-teams=<numTeams>]
                [--map-size=<mapSize>]] [[--visibility=<visibility>] |
                [--terrain-symmetry=<terrainSymmetry> | [--style=<mapStyle> |
                [[--land-density=<landDensity>]
                [--plateau-density=<plateauDensity>]
                [--mountain-density=<mountainDensity>]
                [--ramp-density=<rampDensity>]
                [--reclaim-density=<reclaimDensity>]
                [--mex-density=<mexDensity>] [--biome=<biomeName>]]]]] [COMMAND]
Generates a map from scratch
  -h, --help                 Show this help message and exit.
  -V, --version              Print version information and exit.
      --map-name=<mapName>   Name of map to recreate. Must be of the form
                               neroxis_map_generator_version_seed_options, if
                               present other parameter options will be ignored
      --out-path, --folder-path=<outputPath>
                             Folder to save the map to
      --num-to-generate=<numToGenerate>
                             Number of maps to create
      --seed=<seed>          Seed for the generated map
      --map-size=<mapSize>   Generated map size, can be specified in oGrids (e.
                               g 512) or km (e.g 10km)
                               Default: 512
      --spawn-count=<spawnCount>
                             Spawn count for the generated map
                               Default: 6
      --num-teams=<numTeams> Number of teams for the generated map (0 is no
                               teams asymmetric)
                               Default: 2
      --preview-path=<previewFolder>
                             Folder to save the map previews to
      --debug                Enable debugging
      --visualize            Enable visualization
Options to create a custom map style
      --land-density=<landDensity>
                             Land density for the generated map. Min: 0 Max: 1
      --plateau-density=<plateauDensity>
                             Plateau density for the generated map. Min: 0 Max:
                               1
      --mountain-density=<mountainDensity>
                             Mountain density for the generated map. Min: 0
                               Max: 1
      --ramp-density=<rampDensity>
                             Ramp density for the generated map. Min: 0 Max: 1
      --reclaim-density=<reclaimDensity>
                             Reclaim density for the generated map. Min: 0 Max:
                               1
      --mex-density=<mexDensity>
                             Mex density for the generated map. Min: 0 Max: 1
      --biome=<biomeName>    Texture biome for the generated map. Values:
                               BRIMSTONE, DESERT, EARLYAUTUMN, FRITHEN, MARS,
                               MOONLIGHT, PRAYER, STONES, SUNSET, SYRTIS,
                               WINDINGRIVER, WONDER
Options to specify the map style
      --style=<mapStyle>     Style for the generated map. Values: BASIC,
                               BIG_ISLANDS, CENTER_LAKE, DROP_PLATEAU, FLOODED,
                               HIGH_RECLAIM, LAND_BRIDGE, LITTLE_MOUNTAIN,
                               LOW_MEX, MOUNTAIN_RANGE, ONE_ISLAND,
                               SMALL_ISLANDS, VALLEY
Options to set the generated map visibility
      --visibility=<visibility>
                             Visibility for the generated map. Values:
                               TOURNAMENT, BLIND, UNEXPLORED
Options to set the generated map symmetry
      --terrain-symmetry=<terrainSymmetry>
                             Base terrain symmetry for the generated map.
                               Values: POINT2, POINT3, POINT4, POINT5, POINT6,
                               POINT7, POINT8, POINT9, POINT10, POINT11,
                               POINT12, POINT13, POINT14, POINT15, POINT16, XZ,
                               ZX, X, Z, QUAD, DIAG, NONE
Commands:
  biomes, --biomes  Prints the biomes available
  styles, --styles  Prints the styles available

Comment (BlackYps) [reply generator/src/main/java/com/faforever/neroxis/generator/cli/StyleOptions.java]: I'm not even sure if it makes sense to have a group heading for the last three groups, as there is only one parameter in each group.

File: generator/src/main/java/com/faforever/neroxis/generator/cli/BasicOptions.java (hunk 2 above)
Comment (Sheikah45) [reply generator/src/main/java/com/faforever/neroxis/generator/cli/BasicOptions.java:23]: I reorganized the values and headings

**Comment by Sheikah45** [comment]
I added tests that show that when only terrain symmetry is specified the map is not reproducible from the map name.

File: generator/src/main/java/com/faforever/neroxis/generator/MapGenerator.java [hunk 3]
```diff
@@ -338,54 +216,55 @@ private void verifyNameArgs(String mapName, String[] nameArgs) {
...
+                    generationOptions.getCasualOptions().getStyleOptions().setMapStyle(MapStyle.values()[optionBytes[4]]);
+                } else if (optionBytes.length == 11) {
+                    generatorParametersBuilder.biome(Biomes.loadBiome(BiomeName.values()[optionBytes[4]]));
+                    generatorParametersBuilder.landDensity(MathUtil.normalizeBin(optionBytes[5], NUM_BINS));
+                    generatorParametersBuilder.plateauDensity(MathUtil.normalizeBin(optionBytes[6], NUM_BINS));
+                    generatorParametersBuilder.mountainDensity(MathUtil.normalizeBin(optionBytes[7], NUM_BINS));
+                    generatorParametersBuilder.rampDensity(MathUtil.normalizeBin(optionBytes[8], NUM_BINS));
+                    generatorParametersBuilder.reclaimDensity(MathUtil.normalizeBin(optionBytes[9], NUM_BINS));
+                    generatorParametersBuilder.mexDensity(MathUtil.normalizeBin(optionBytes[10], NUM_BINS));
+                }
```
Comment (Sheikah45) [inline generator/src/main/java/com/faforever/neroxis/generator/MapGenerator.java]: Collapse the else {  if { to just else if {

**Review by Sheikah45 — CHANGES_REQUESTED** [review]

**Review by Sheikah45 — APPROVED** [review]

**Pull request merged**